pkg/config/viper.go

```diff
@@ -8,66 +8,251 @@ package config
 import (
 	"io"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
 )
 
-// Config represents an object that can load and store configuration parameters
-// coming from different kind of sources:
-// - defaults
-// - files
-// - environment variables
-// - flags
-type Config interface {
-	Set(key string, value interface{})
-	SetDefault(key string, value interface{})
-	IsSet(key string) bool
-
-	Get(key string) interface{}
-	GetString(key string) string
-	GetBool(key string) bool
-	GetInt(key string) int
-	GetInt64(key string) int64
-	GetFloat64(key string) float64
-	GetTime(key string) time.Time
-	GetDuration(key string) time.Duration
-	GetStringSlice(key string) []string
-	GetStringMap(key string) map[string]interface{}
-	GetStringMapString(key string) map[string]string
-	GetStringMapStringSlice(key string) map[string][]string
-	GetSizeInBytes(key string) uint
-
-	SetEnvPrefix(in string)
-	BindEnv(input ...string) error
-	SetEnvKeyReplacer(r *strings.Replacer)
-
-	UnmarshalKey(key string, rawVal interface{}) error
-	Unmarshal(rawVal interface{}) error
-	UnmarshalExact(rawVal interface{}) error
-
-	ReadInConfig() error
-	ReadConfig(in io.Reader) error
-	MergeConfig(in io.Reader) error
-
-	AllSettings() map[string]interface{}
-
-	AddConfigPath(in string)
-	SetConfigName(in string)
-	SetConfigFile(in string)
-	SetConfigType(in string)
-	ConfigFileUsed() string
-
-	BindPFlag(key string, flag *pflag.Flag) error
+// safeConfig wraps viper with a safety lock
+type safeConfig struct {
+	*viper.Viper
+	sync.RWMutex
+}
+
+// Set is wrapped for concurrent access
+func (c *safeConfig) Set(key string, value interface{}) {
+	c.Lock()
+	c.Viper.Set(key, value)
+	c.Unlock()
```

Comment:
I think we should either use `defer` everywhere or none anywhere. Just to keep it consistent. If this API is not heavily called then I think the minor gain from having no `defer` is not that important. I'd be fine with any.